### ISSUE TITLE
Behaviour of unhandled prompt handler diverges from Selenium

### ISSUE THREAD

**Issue by shs96c** [w3c/webdriver#912]
In Selenium, setting the unhandled prompt handler behaves in the same way as in the spec, but with the important difference that if a prompt is present, an error will always be returned. In the spec, we return `success` if the handler is invoked.

The intention is to allow a user to re-use their existing WebDriver instance without needing to create a new one (which can be a slow and expensive process), but they should still be notified that the driver is in an unexpected state. It would seem that the Selenium behaviour better meets that intention, and I suspect that the [wording in the user prompts section](https://w3c.github.io/webdriver/webdriver-spec.html#dfn-handle-any-user-prompts) is an oversight.

Test for Selenium behaviour:
https://github.com/SeleniumHQ/selenium/blob/master/dotnet/test/common/AlertsTest.cs#L681

Existing implementation in the legacy FirefoxDriver (all code paths lead to an error being returned to the user if an unexpected modal is open):
https://github.com/SeleniumHQ/selenium/blob/master/javascript/firefox-driver/js/nsCommandProcessor.js#L478
https://github.com/SeleniumHQ/selenium/blob/master/javascript/firefox-driver/js/modals.js#L261

**Comment by shs96c** [comment]
Relevant Selenium issue: https://github.com/SeleniumHQ/selenium/issues/3972

**Comment by andreastt** [comment]
> The intention is to allow a user to re-use their existing WebDriver instance without needing to create a new one (which can be a slow and expensive process), but they should still be notified that the driver is in an unexpected state.

The driver doesn’t get into an unrecoverable state when a user prompt is present. You can call the Dismiss Alert command to close any prompt if a command returns with an ‘unexpected alert open’ error, so I don’t understand why a user would have to create a new session?

> It would seem that the Selenium behaviour better meets that intention, and I suspect that the wording in the user prompts section is an oversight.

From what you’re writing, I don’t understand what behaviour you’re seeking.

What I find mildly surprising about the user prompt handler as it is defined right now, is that the prompt is dismissed if a handler is not set. I would only expect it to dismiss a prompt if the handler is in the "dismiss" state. In the default/unknown state, I would expect it to just return an error.

**Comment by shs96c** [comment]
Consider these tests, which share a common WebDriver instance (written in the Java style, but the intent should be clear):

```
@Test
public void failsBecauseOfAnUnexpectedAlert() {
  driver.get("http://www.example.com/");
  WebElement element = driver.findElement(By.id("cake"));
  element.click();  // This causes the alert to happen
  assertEquals("cheese", cake.getText());
}

@Test
public void runImmediatelyAfterwardsWithTheSameDriver() {
  driver.get("http://www.example.com/some-page");
}
```

Here, the user isn't even aware of the fact that an unexpected user prompt has been displayed. The assertion on the first test will have caused _something_ to happen. What is that "something"?

In existing selenium tests, we'd act upon the unhandled prompt handler, and then throw an exception detailing that an unexpected user prompt was present, and the text of that prompt. 

Following the spec, if there is no handler, we default to the existing selenium behaviour. If there is a handler, the assertion will never cause the user to be notified that something untoward has happened.

So the argument isn't that the driver is in an unrecoverable state, it's that:

- the user will lose context and data that are important in the common use case of testing. 
- the user will simply have a failing test and will be unaware that they need to close the alert manually.

In either case, the default behaviour we have in the spec is to support users who aren't aware of the fact that a test might have opened an alert. This was the original reason for adding this behaviour --- indeed, Selenium's own test suites would fail horribly if there was an unexpected user prompt, as we don't check for the presence of an `Alert` and attempt to close it before continuing our tests. There's also no method in the spec that allows a user to detect whether or not an alert is present without an error being returned if there is no user prompt present, so you'd have to rely on exceptions for flow control, which is seldom an advisable programming practice.

You could argue at this point that users will only set the unhandled prompt handler if they're aware of this change in behaviour. Perhaps we need `"accept and continue"` and `"dismiss and continue"` to cover all the bases?

**Comment by shs96c** [comment]
Closed by #996
